specification.md

```diff
@@ -480,7 +480,7 @@ into. With this, after our action executes the state data would be:
 
 | Parameter | Description | Type | Required |
 | --- | --- | --- | --- |
-| data | Workflow expression that filters of the event data (payload) | string | no |
+| data | Workflow expression that filters the event data (payload) | string | no |
```

Comment:
Does this imply that they can only access the 'data' portion of the event and not other attributes of the event?

Reply:
the context attributes of the CE are to be used when you define the event (event definition) and when you define your correlations. When state consumes that defined event we are dealing with its data (payload) only and we require it to be json format. given that the CE event itself (context attributes) are not always json format I think that is reasonable. wdyt?

Reply:
Just a brief skim of CE, the first thing I notice is an ID attribute that is a top-level attribute of the CE.  It's supposed to uniquely identify the event.  With this behavior, it's impossible for a workflow to reference that ID, right?

The subject and time attributes are other top-level attributes of a CE that seem like they might be useful for a workflow.

Reply:
the event that you consume is defined in the event definition and the correlation. 
a CE over http can look like this:

POST /someresource HTTP/1.1
Host: webhook.example.com
ce-specversion: 1.0
ce-type: com.example.someevent
ce-time: 2018-04-05T03:56:24Z
ce-id: 1234-1234-1234
ce-source: /mycontext/subcontext
    .... further attributes ...
Content-Type: application/json; charset=utf-8
Content-Length: nnnn

{
    ... application data ...
}

Reply:
if you want those data to be available put them in the payload of the event that is json

Reply:
That doesn't change the fact that these are attributes of the cloud event.  If a runtime is accepting cloud-events in this format and thus already knows how to get type & source from the headers to do correlation, then it knows enough to get and present the entire cloud-event to the workflow as this instead:

```
{
  specversion: 1.0,
  type: com.example.someevent,
  time: 2018-04-05T03:56:24Z,
  id: 1234-1234-1234,
  source: /mycontext/subcontext,
  "data": {
    ... application data ...
  }
}
```

I'm not sure why we should force users of serverless workflow to alter the structure of their events just to integrate serverless workflows into their existing cloud-event use-cases.

Reply:
That is just an example given in the CE spec doc that is written in json. CE has a number of defined protocol bindings, 
https://github.com/cloudevents/spec
see the *-protocol-bindings.md documents

we can enforce CE for defining what events workflow needs with the required context attributes, but we cannot enforce that the CE message is in json format. 
the runtimes typically get the entire event (whatever protocol/format it comes in) and can add all the relevant information, but from our point of view how can we assure that every single one does it?

Reply:
there are limitations ..."data" being json is pretty much a standard, but we cannot for example use the 
https://github.com/cloudevents/spec/blob/v1.0.1/spec.json#L75
element which is a shame

Reply:
If we allowed access to the full cloud-event, then actually we could use that, couldn't we?  User could pass the following as an argument to a function, for example: `${ .data_base64 }`

I actually think this implies that we get better integration w/ cloud-events if we give the user access to the full set of all attributes of the cloud-event.

> but we cannot enforce that the CE message is in json format

I agree that we can't enforce what wire-format the runtime receives the cloud-event in.  But I don't see why we can't say that the runtime should present any received event in the json form to the workflow.

I understand that cloud-event has lots of different protocol bindings, but afaik, none of them stop us from presenting the cloud-event as the json form of a cloud-event.

Reply:
@jorgenj ok. i will try to get some CE guy in one of our meetings so we can ask them all these questions. would it be ok if we leave it as it is for now (just data) as that part did not change, and raise an issue to change this across the board if we want soon?

Reply:
Created an issue: #284